lib/src/boundimage.rs

```diff
@@ -0,0 +1,171 @@
+            let entry = entry?;
+            let file_name = entry.file_name();
+            let file_name = if let Some(n) = file_name.to_str() {
+                n
+            } else {
+                anyhow::bail!(
+                    "Invalid non-UTF8 filename: {file_name:?} in {}",
+                    &self.spec_dir
+                );
+            };
+
+            if !entry.file_type()?.is_symlink() {
+                anyhow::bail!("Not a symlink: {file_name}");
+            }
+
+            //parse the file contents
+            let file_path = entry.path();
+            let file_path = file_path.strip_prefix(format!("/{}", &self.deployment_root))?;
+
+            let root_dir = File::open(&self.deployment_root)?;
+            let root_fd = root_dir.as_fd();
+
+            let fd = rustix::fs::openat2(
+                root_fd,
+                file_path,
+                OFlags::empty(),
+                rustix::fs::Mode::empty(),
+                ResolveFlags::IN_ROOT,
+            )?;
+
+            let mut file = { File::from(fd) };
+            let mut file_contents = String::new();
+            file.read_to_string(&mut file_contents)?;
+
+            let file_ini = ini::Ini::load_from_str(&file_contents).context("Parse to ini")?;
+            let file_extension = Utf8Path::new(file_name).extension();
+            let bound_image = match file_extension {
+                Some("image") => self.parse_image_file(file_name, &file_ini),
+                Some("container") => self.parse_container_file(file_name, &file_ini),
+                _ => anyhow::bail!("Invalid file extension: {file_name}"),
+            }?;
+
+            bound_images.push(bound_image);
+        }
+
+        Ok(bound_images)
+    }
+
+    #[context("parse image file {file_name}")]
+    fn parse_image_file(&self, file_name: &str, file_contents: &ini::Ini) -> Result<BoundImage> {
+        let image = if let Some(img) = file_contents.get_from(Some("Image"), "Image") {
+            img
+        } else {
+            anyhow::bail!("Missing Image field in {file_name}");
+        };
+
+        let auth_file = file_contents
+            .get_from(Some("Image"), "AuthFile")
+            .map(|s| s.to_string());
+
+        let bound_image = BoundImage::new(image.to_string(), auth_file)?;
+        Ok(bound_image)
+    }
+
+    #[context("parse container file {file_name}")]
+    fn parse_container_file(
+        &self,
+        file_name: &str,
+        file_contents: &ini::Ini,
+    ) -> Result<BoundImage> {
+        let image = if let Some(img) = file_contents.get_from(Some("Container"), "Image") {
+            img
+        } else {
+            anyhow::bail!("Missing Image field in {file_name}");
+        };
+
+        let bound_image = BoundImage::new(image.to_string(), None)?;
+        Ok(bound_image)
+    }
+
+    #[context("pull bound images")]
+    fn pull_images(&self, bound_images: Vec<BoundImage>) -> Result<()> {
+        //TODO: do this in parallel
```

Comment:
If we did do that it'd open up interesting questions around how we display progress bars. Are we doing that today via podman? I think the `Task` bits may default stdin to `/dev/null`? If we're not then we should probably explicitly provide e.g. `-q`.

One thing I'm thinking here is perhaps we go ahead and generate a systemd unit for each container to pull (exactly like quadlet would do)?

But, clearly this is not a blocker for landing this PoC work.